rust/xaynet-mobile/src/participant.rs

```diff
@@ -0,0 +1,239 @@
+        let client = Client::new(url, None, None).map_err(InitError::Client)?;
+        let (events, notifier) = Events::new();
+        let store = Store::new();
+        let state_machine = StateMachine::new(pet_settings, client, store.clone(), notifier);
+        Self::init(state_machine, events, store)
+    }
+
+    pub fn restore(state: &[u8], url: String) -> Result<Self, InitError> {
+        let state: SerializableState =
+            bincode::deserialize(state).map_err(InitError::Deserialization)?;
+        let (events, notifier) = Events::new();
+        let store = Store::new();
+        let client = Client::new(url, None, None).map_err(InitError::Client)?;
+        let state_machine = StateMachine::restore(state, client, store.clone(), notifier);
+        Self::init(state_machine, events, store)
+    }
+
+    fn init(state_machine: StateMachine, events: Events, store: Store) -> Result<Self, InitError> {
+        let mut participant = Self {
+            runtime: Self::runtime()?,
+            state_machine: Some(state_machine),
+            events,
+            store,
+            task: Task::None,
+            made_progress: true,
+            should_set_model: false,
+        };
+        participant.process_events();
+        Ok(participant)
+    }
+
+    fn runtime() -> Result<Runtime, InitError> {
+        tokio::runtime::Builder::new()
+            .basic_scheduler()
+            .enable_all()
+            .build()
+            .map_err(InitError::Runtime)
+    }
+
+    pub fn save(self) -> Vec<u8> {
+        // UNWRAP_SAFE: the state machine is always set.
+        let state_machine = self.state_machine.unwrap().save();
+        bincode::serialize(&state_machine).unwrap()
+    }
+
+    pub fn tick(&mut self) {
+        // UNWRAP_SAFE: the state machine is always set.
+        let state_machine = self.state_machine.take().unwrap();
+        let outcome = self
+            .runtime
+            .block_on(async { state_machine.transition().await });
+        match outcome {
+            TransitionOutcome::Pending(new_state_machine) => {
+                self.made_progress = false;
+                self.state_machine = Some(new_state_machine);
+            }
+            TransitionOutcome::Complete(new_state_machine) => {
+                self.made_progress = true;
+                self.state_machine = Some(new_state_machine)
+            }
+        };
+        self.process_events();
+    }
+
+    fn process_events(&mut self) {
+        loop {
+            match self.events.next() {
+                Some(Event::Idle) => {
+                    self.task = Task::None;
+                }
+                Some(Event::Update) => {
+                    self.task = Task::Update;
+                }
+                Some(Event::Sum) => {
+                    self.task = Task::Sum;
+                }
+                // not sure whether we need to do anything here
+                Some(Event::NewRound) => {}
```

Comment:
> Not sure if I followed the code correctly but I think we will need to update the task otherwise the match in xaynet_ffi_participant_tick returns the old mask

You really reviewed this carefully to spot this ^^ It should be taken care of because state machine is supposed to emit an "idle" notification when a new round starts. But that requires tests, both in `xaynet_sdk` to make sure that this event is indeed emitted, and here.

> we can notify the user of the client that a new global model is available I think this method is currently missing in the ffi part.

Indeed. But if you don't mind I'd prefer to do this in a follow-up PR because it's not _strictly_ necessary.

Reply:
> You really reviewed this carefully to spot this ^^ It should be taken care of because the state machine is supposed to emit an "idle" notification when a new round starts. But that requires tests, both in xaynet_sdk to make sure that this event is indeed emitted, and here.

true I looked at the current master sry😅
> Indeed. But if you don't mind I'd prefer to do this in a follow-up PR because it's not strictly necessary.

fine for me